blazar/manager/service.py

```diff
@@ -770,7 +770,37 @@ def _allocation_candidates(self, lease, reservations):
                 raise common_ex.BlazarException(
                     'Invalid plugin names are specified: %s' % resource_type)
 
-            candidate_ids = plugin.allocation_candidates(res)
+            try:
+                if can_retry:
+                    plugin.update_default_parameters(res)
+                candidate_ids = plugin.allocation_candidates(res)
+            except exceptions.NotEnoughResourcesAvailable as ex:
+                # If the plugin doesn't support default resource properties
+                # raise the exception
+                if not hasattr(
+                    CONF[plugin.resource_type],
+                    "retry_allocation_without_defaults"
+                ) or not hasattr(
+                    CONF[plugin.resource_type],
+                    "display_default_resource_properties"
+                ):
+                    raise ex
+
+                # Retry this function if allowed
+                if (CONF[plugin.resource_type]
+                        .retry_allocation_without_defaults and can_retry):
+                    LOG.info("RETRYING CANDIDATES")
+                    return self._allocation_candidates(
```

Comment:
Is it possible to do the same w/o having to re-run the entire function? It seems like the main thing is trying to get a new list of allocation candidates. Would this work? I find the "retry" mechanism a bit difficult to parse and am wondering if it can be simpler.

```python
original_res = res.copy()  # deep-copy?
try:
    plugin.update_default_parameters(res)
    candidate_ids = plugin.allocation_candidates(res)
except exceptions.NotEnoughResourcesAvailable as ex:
    # Check if we _should_ retry given config, then...
    LOG.info("Not enough resources with default properties. Retrying with defaults removed.")
    candidate_ids = plugin.allocation_candidates(original_res)
```